### PR TITLE
GroundStation HMAC Support

### PR DESCRIPTION
**Pull request by dmitri-mcguckin**
# Features
* Add post-processing-hook step to generate and append an HMAC key to the payload
* Integrated Yamcs LoggingHandler + relevant configs into EdlPostProcessing hook
* Reduced maximum file-name size from 32 bytes to 8 bytes

# Bug Fixes
* Replaced stale reference to `encodeShort` with `encodeUnsignedShort` which was crashing Yamcs in Java 11

# To Do
* Update secret loading to come from either the Yamcs central config, or an environment variable or something
* Update syslogger.properties to point to `/var/log` instead

### CODE REVIEW

**Review by ryanpdx — COMMENTED** [review]
When I tried to run it  got:
```
[INFO] --- yamcs-maven-plugin:1.2.7:run (default-cli) @ uniclogs-yamcs ---
[INFO] Creating configuration at /home/ryan/PSAS/repos/uniclogs-yamcs/target/yamcs
Can't load log handler "java.util.logging.FileHandler"
java.nio.file.NoSuchFileException: /opt/yamcs/log/yamcs-server.log.0.lck
java.nio.file.NoSuchFileException: /opt/yamcs/log/yamcs-server.log.0.lck
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:182)
	at java.base/java.nio.channels.FileChannel.open(FileChannel.java:298)
	at java.base/java.nio.channels.FileChannel.open(FileChannel.java:357)
	at java.logging/java.util.logging.FileHandler.openFiles(FileHandler.java:512)
	at java.logging/java.util.logging.FileHandler.<init>(FileHandler.java:279)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:128)
	at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:347)
	at java.base/java.lang.Class.newInstance(Class.java:645)
	at java.logging/java.util.logging.LogManager.createLoggerHandlers(LogManager.java:1005)
	at java.logging/java.util.logging.LogManager$4.run(LogManager.java:975)
	at java.logging/java.util.logging.LogManager$4.run(LogManager.java:971)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at java.logging/java.util.logging.LogManager.loadLoggerHandlers(LogManager.java:971)
	at java.logging/java.util.logging.LogManager.initializeGlobalHandlers(LogManager.java:2424)
	at java.logging/java.util.logging.LogManager$RootLogger.accessCheckedHandlers(LogManager.java:2526)
	at java.logging/java.util.logging.Logger.getHandlers(Logger.java:2090)
	at java.logging/java.util.logging.Logger.log(Logger.java:977)
	at org.yamcs.logging.Log.log(Log.java:147)
	at org.yamcs.logging.Log.info(Log.java:50)
	at org.yamcs.YamcsServer.setupLogging(YamcsServer.java:1003)
	at org.yamcs.YamcsServer.main(YamcsServer.java:926)
Feb 18, 2022 9:45:53 PM org.yamcs.logging.Log log
INFO: Logging enabled using /home/ryan/PSAS/repos/uniclogs-yamcs/target/yamcs/etc/logging.properties
Feb 18, 2022 9:45:54 PM org.yamcs.logging.Log log
INFO: Loading service HttpServer
Feb 18, 2022 9:45:54 PM org.yamcs.logging.Log log
INFO: Creating or loading tablespace _global
Feb 18, 2022 9:45:54 PM org.yamcs.logging.Log log
INFO: Opened tablespace database /home/ryan/PSAS/repos/uniclogs-yamcs/target/yamcs/yamcs-data/_global.rdb, num records:6, num metadata records: 4, maxTbsIndex: 4
Feb 18, 2022 9:45:54 PM org.yamcs.logging.Log log
INFO: Tablespace _global: loaded 0 tables for instance _global
Feb 18, 2022 9:45:54 PM org.yamcs.logging.Log log
INFO: Loading online instance 'oresat0'
Feb 18, 2022 9:45:54 PM org.yamcs.xtce.xml.XtceStaxReader readXmlDocument
INFO: Parsing XTCE file mdb/oresat0-xtce.xml
Feb 18, 2022 9:45:54 PM org.yamcs.xtce.xml.XtceStaxReader skipXtceSection
INFO: Section <AuthorSet> skipped
Feb 18, 2022 9:45:54 PM org.yamcs.xtce.xml.XtceStaxReader readXmlDocument
INFO: XTCE file parsing finished, loaded: 132 parameters, 1 tm containers, 28 commands
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Serialized database saved locally
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Creating or loading tablespace oresat0
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Opened tablespace database /home/ryan/PSAS/repos/uniclogs-yamcs/target/yamcs/yamcs-data/oresat0.rdb, num records:10449, num metadata records: 41, maxTbsIndex: 39
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Tablespace oresat0: loaded 6 tables for instance oresat0
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading service XtceTmRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading service ParameterRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading service AlarmRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading service EventRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading service ReplayServer
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading service SystemParametersService
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading service ProcessorCreatorService
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading service CommandHistoryRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading service ParameterArchive
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading processor service StreamTmPacketProvider
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading processor service StreamTcCommandReleaser
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading processor service StreamParameterProvider
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading processor service AlgorithmManager
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Loading processor service LocalParameterManager
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Creating new processor 'realtime' of type 'realtime'
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: yamcs 5.5.7, build 39b6a401
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Awaiting start of service XtceTmRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Awaiting start of service ParameterRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Awaiting start of service AlarmRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Awaiting start of service EventRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Awaiting start of service ReplayServer
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Awaiting start of service SystemParametersService
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Awaiting start of service ProcessorCreatorService
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Awaiting start of service CommandHistoryRecorder
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Awaiting start of service ParameterArchive
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Yamcs started in 2127ms. Started 1 of 1 instances and 10 services
Feb 18, 2022 9:45:55 PM org.yamcs.logging.Log log
INFO: Website deployed at http://razer:8090
````

That is a lot of logging. 
Also, i don't think it good idea assume the program can write logs to `/opt/`.

**Comment by ryanpdx** [comment]
I guess we can target this project to be for docker containers only.  That would make the logging to `/opts/` error go away. Also, this would allow us to configure yamcs to use persistent data (requires a absolute path (see comment `src/main/yamcs/etc/yamcs.yaml`).

**Comment by dmitri-mcguckin** [comment]
I think all you're missing is a bit of environment contex.

The logs happen to be in /opt because that's where my target install is at the moment.

The point is just that it's in the same directory as your install path.

Feel free to do so for your own and it should work then.

Though we could probably also move the log output to /var/log if we wanted to future proof some more.

The goal of this pr is not future proofing though, it's to get something working and ready for link testing asap.

**Comment by dmitri-mcguckin** [comment]
I don't think a docker container is strictly necessary, definitely not right now, because trying to support changing the xtce's and such and then all of that inside an image is more complexity we shouldn't add until yamcs is working.

**Comment by ryanpdx** [comment]
Okay, sounds good.

We should have the default settings log to the current working dir not `/opt/` then, so it can run on anyones computer.

File: src/main/yamcs/mdb/oresat0-xtce.xml [hunk 1]
```diff
@@ -628,11 +628,13 @@
             <xtce:MetaCommand name="C3FwBank" shortDescription="swap firmware bank">
                 <xtce:ArgumentList>
                     <xtce:Argument name="Bank" argumentTypeRef="Uint32Type"/>
+                    <xtce:Argument name="HMAC32" argumentTypeRef="Uint32Type"/>
                 </xtce:ArgumentList>
                 <xtce:CommandContainer name="C3FwBank">
                     <xtce:EntryList>
                         <xtce:FixedValueEntry binaryValue="02" sizeInBits="8"/>
                         <xtce:ArgumentRefEntry argumentRef="Bank"/>
+                        <xtce:ArgumentRefEntry argumentRef="HMAC32"/>
```
Comment (ryanpdx) [inline src/main/yamcs/mdb/oresat0-xtce.xml]: What is the HMAC32 arg for? Isn't this added in the post processor?

Comment (dmitri-mcguckin) [reply src/main/yamcs/mdb/oresat0-xtce.xml]: Hmn, I thought I had removed that before. That predates our discussion on the post-processor.

File: src/main/yamcs/etc/logging.properties [hunk 2]
```diff
@@ -0,0 +1,12 @@
+handlers = java.util.logging.ConsoleHandler, java.util.logging.FileHandler
+
+java.util.logging.ConsoleHandler.level = INFO
+
+
+java.util.logging.FileHandler.level = ALL
+java.util.logging.FileHandler.formatter = org.yamcs.logging.CompactFormatter
+java.util.logging.FileHandler.pattern = /opt/yamcs/log/yamcs-server.log
```
Comment (ryanpdx) [inline src/main/yamcs/etc/logging.properties]: Should not assume Yamcs has permission to write to `/opt/`.

**Review by ryanpdx — CHANGES_REQUESTED** [review]

File: src/main/yamcs/etc/logging.properties (hunk 2 above)
Comment (dmitri-mcguckin) [reply src/main/yamcs/etc/logging.properties]: There's no elegant way to make it write to a permission-safe directory without running yamcs as root or running a script as root temporarily in order to make and change the owners of a dir.

Right now, this does not crash yamcs, it just stops the file-based logging (still does console logging) if it happens to not be installed in `/opt`.

Once again, right now is not polish time and future proofing, this is a minimum viable product.

We can spend time later, looking at more elegant ways to ensure the log directory exists (probably via maven somehow) but this isn't a road blocker.

**Review by ryanpdx — DISMISSED** [review]

**Pull request merged**